File: python/addsUpToK.py
def all_pairs_that_add_up_to_k(array, k):
    seen_numbers = set()
    pairs = []

    for num in array:
        complement = k - num
        if complement in seen_numbers:
            pairs.append((complement, num))
        seen_numbers.add(num)

    return pairs if pairs else None

# Test the function
#test1
# arr = [10, 15, 3, 7, 7, 10]
# knumber = 17
#test2
arr = [10, 15, 3, 7, 6, 10, 4, 7]
knumber = 14
result = all_pairs_that_add_up_to_k(arr, knumber)
if result:
    print(f"Pairs that add up to {knumber}: {result}")
else:
    print(f"No pairs add up to {knumber}")

File: go/addsUpToK.go
package main

import (
	"fmt"
)

func allPairsThatAddUpToK(array []int, k int) [][]int {
	seenNumbers := make(map[int]bool)
	pairs := [][]int{}

	for _, num := range array {
		complement := k - num
		if seenNumbers[complement] {
			pairs = append(pairs, []int{complement, num})
		}
		seenNumbers[num] = true
	}

	if len(pairs) == 0 {
		return nil
	}
	return pairs
}

func main() {
	arr := []int{10, 15, 3, 7, 7, 10}
	k := 17
	result := allPairsThatAddUpToK(arr, k)

	if result != nil {
		fmt.Printf("Pairs that add up to %d: %v\n", k, result)
	} else {
		fmt.Printf("No pairs add up to %d\n", k)
	}
}
